src/s2wasm.h

```diff
@@ -229,29 +229,68 @@ class S2WasmBuilder {
   // gets a constant, which may be a relocation for later.
   // returns whether this is a relocation
   // TODO: Clean up this and the way relocs are created from parsed objects
-  bool getRelocatableConst(uint32_t* target) {
+  LinkerObject::Relocation* getRelocatableConst(uint32_t* target) {
     if (isdigit(*s) || *s == '-') {
       int32_t val = getInt();
       memcpy(target, &val, sizeof(val));
-      return false;
-    } else {
-      // a global constant, we need to fix it up later
-      Name name = getStrToSep();
-      LinkerObject::Relocation::Kind kind = isFunctionName(name) ?
-          LinkerObject::Relocation::kFunction :
-          LinkerObject::Relocation::kData;
-      int offset = 0;
-      if (*s == '+') {
-        s++;
-        offset = getInt();
-      } else if (*s == '-') {
-        s++;
-        offset = -getInt();
-      }
-      linkerObj->addRelocation(kind, target,
-                               fixEmLongjmp(cleanFunction(name)), offset);
-      return true;
+      return nullptr;
+    }
+
+    // a global constant, we need to fix it up later
+    Name name = getStrToSep();
+    LinkerObject::Relocation::Kind kind = isFunctionName(name) ?
+        LinkerObject::Relocation::kFunction :
+        LinkerObject::Relocation::kData;
+    int offset = 0;
+    if (*s == '+') {
+      s++;
+      offset = getInt();
+    } else if (*s == '-') {
+      s++;
+      offset = -getInt();
+    }
+    return new LinkerObject::Relocation(
+        kind, target, fixEmLongjmp(cleanFunction(name)), offset);
+  }
+  Expression* relocationToGetGlobal(LinkerObject::Relocation* relocation) {
+    if (!relocation) {
+      return nullptr;
+    }
+
+    auto name = relocation->symbol;
+    auto g = allocator->alloc<GetGlobal>();
+    g->name = name;
+    g->type = i32;
+
+    // Optimization: store any nonnegative addends in their natural place.
+    // Only do this for positive addends because load/store offsets cannot be
+    // negative.
+    if (relocation->addend >= 0) {
+      *relocation->data = relocation->addend;
```

Comment:
Trivia: there's an interesting, annoyingly-subtle interaction here.
There's four places that call `getRelocatableConst`: load, store, const, and data section initializer.
In data, we don't model `get_global` at all, so we abort here for now.
For load and store, this writes to the `offset=___` field, and is why we do this.
Const is the odd one, because the pointer we write to is the value of the const we're creating. Without this that would become an `(i32.const 0)`, that the other optimization would remove.
Either way we turn `i32.const $push0=, foo+2` into `(i32.add (i32.const 2) (get_global $foo))`

I imagine this is quirky enough that it will someday become a problem, and it's probably worth just emitting a bunch of `(i32.add (i32.const 0)`s and letting the optimizer take care of it, just for sanity's sake.